### ISSUE TITLE
GPU rendering seems to have stopped working for me

### ISSUE THREAD

**Issue by thennicke** [jliljebl/flowblade#1135]
1) Flowblade version: 2.10.0.4
2) MLT version: 7.17.0
3) Your distribution: Fedora

Output of `vainfo`:
```
Trying display: wayland
libva info: VA-API version 1.18.0
libva info: Trying to open /usr/lib64/dri/radeonsi_drv_video.so
libva info: Found init function __vaDriverInit_1_18
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.18 (libva 2.18.2)
vainfo: Driver version: Mesa Gallium driver 23.1.9 for AMD Radeon RX 6800 (navi21, LLVM 16.0.6, DRM 3.54, 6.5.10-200.fc38.x86_64)
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileVP9Profile0            :	VAEntrypointVLD
      VAProfileVP9Profile2            :	VAEntrypointVLD
      VAProfileAV1Profile0            :	VAEntrypointVLD
      VAProfileNone                   :	VAEntrypointVideoProc
```

My GPU was working before; I was able to render 1080p30 videos at about 2x the playback speed (using the `VAAPI h.246 / mp4` output format). I don't know what has changed but the rendering is using just the CPU now and is about 4-8 times slower. I'm getting the following line in terminal but don't know how to interpret the codes:

```
GPU test results {'NVENC H.264 High Profile / .mp4': -11, 'NVENC HEVC Main10 Profile / .mp4': -11, 'VAAPI H.264 / .mp4': 0}
```
What does `0` mean? What does `-11` mean? Am I set up correctly? Any recommendations?

Thanks very much

**Comment by jliljebl** [comment]
Unfortunately I have experienced this this too after updating to Ubuntu 23.10. The reason in my view is that *ffmpeg* is no longer compiled with GPU support included. My Flatpak compile flags for ffmpeg are:

```
      - --enable-gpl
      - --disable-static
      - --enable-shared
      - --disable-doc
      - --enable-libfontconfig
      - --enable-libfreetype
      - --enable-libopus
      - --enable-librsvg
      - --enable-libvpx
      - --enable-libx264
      - --enable-libx265
      - --enable-libpulse
      - --enable-libmp3lame
      - --enable-libvorbis
      - --enable-libaom
      - --enable-vaapi
      - --enable-ffnvcodec
      - --enable-nvenc
      - --enable-amf
      - --enable-libzimg
```
For GPU encoding the important flags are: *--enable-vaapi, --enable-ffnvcodec, --enable-nvenc*

However, when running Flowblade with current Ubuntu packaged libraries I get:

```
ffmpeg version 6.0-6ubuntu1 Copyright (c) 2000-2023 the FFmpeg developers
built with gcc 13 (Ubuntu 13.2.0-2ubuntu1)
configuration: --prefix=/usr --extra-version=6ubuntu1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libglslang --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librist --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --disable-sndio --enable-libjxl --enable-pocketsphinx --enable-librsvg --enable-libvpl --disable-libmfx --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libx264 --enable-libplacebo --enable-librav1e --enable-shared
```

which is missing support for either VAAPi or NVENC  because *--enable-vaapi, --enable-ffnvcodec, --enable-nvenc* are not present. Probably this is something that has changed for *ffmpeg version 6.0* as FLATPAK uses version *ffmpeg version 5.1.1*. 

> GPU test results {'NVENC H.264 High Profile / .mp4': -11, 'NVENC HEVC Main10 Profile / .mp4': -11, 'VAAPI H.264 / .mp4': 0}

> What does 0 mean? What does -11 mean? Am I set up correctly? Any recommendations?

Here "0" means "everything succeeded" whereas any other number means that the rendering test failed.  So VAAPI rendering should still work, but you unfortunately you report that it is slower, so the test succeeds because some CPU backup path is used. 

This just one of those things when upstream breaks stuff and we need to discover fixes which are not obvious immediately. This will get fixed, but I cannot give estimates on when at the moment. I'll start doing tests with ffmpeg 6.0 after I get the 2.12 out hopefully next week.

**Comment by thennicke** [comment]
Thank you for the explanation; I'm glad it's not just me with the issue. I will await any updates.
~ Ted

**Comment by jliljebl** [comment]
Ok, I got NVENC rendering working on Ubuntu 23.10. Both a driver update and a Flowblade code change were required.

On a test *ffmpeg* render I was informed that some API version 12 was required and my system had 11.1. Terminal output also gave a minimum NVidia driver version that was required.

I managed to fix this issue by updating drivers with command: 
```
sudo ubuntu-drivers autoinstall
```
In addition I needed to add *ffmpeg* arg *b_ref_mode=0* to encoding options using vcodec *hevc_nvenc*. This was very likely a change made necessary by *fmpeg* version update to 6.0.

Problems with VAAPI encoding probably have similar causes; drivers may need to be updated and some changes in render args may be required.

It would help if someone with VAAPI system (AMD or Intel card) would test the current repository code, and after startup open file *$HOME/.cache/flowblade/log_gputest_render_VAAPIH* and post the last lines with the *ffmpeg* error information here.

**Comment by jliljebl** [comment]
> Trying display: wayland
libva info: VA-API version 1.18.0
libva info: Trying to open /usr/lib64/dri/radeonsi_drv_video.so
libva info: Found init function __vaDriverInit_1_18
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.18 (libva 2.18.2)
vainfo: Driver version: Mesa Gallium driver 23.1.9 for AMD Radeon RX 6800 (navi21, LLVM 16.0.6, DRM 3.54, 6.5.10-200.fc38.x86_64)
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileVP9Profile0            :	VAEntrypointVLD
      VAProfileVP9Profile2            :	VAEntrypointVLD
      VAProfileAV1Profile0            :	VAEntrypointVLD
      VAProfileNone                   :	VAEntrypointVideoProc

Ok, I found the answer here: https://wiki.archlinux.org/title/Hardware_video_acceleration#AMD/ATI

From the wiki:

> VAEntrypointVLD means that your card is capable to decode this format, VAEntrypointEncSlice means that you can encode to this format.

So according to this info your card is not capable of hardware encoding and should not have worked ever

**Comment by jliljebl** [comment]
I'm closing this as *fixed*. We got the NVENC working with ffmpeg version 6.0, and the VAAPI issue is caused by user card  not support GPU encoding.

**Comment by thennicke** [comment]
Oh I had no idea the RX 6800 didn't support GPU encoding. Now that you mentioned it, I replaced my old RX 580 card with this new one; perhaps the old RX 580 supported GPU encoding and this new one doesn't, which would explain why I'm getting different results now. My memory was playing tricks on me -- thanks for resolving this.

**Comment by jliljebl** [comment]
@thennicke Yeah, I bought a RX 580 to be able to test VAAPI encoding but my mobo didn't provide the needed additional power (PCI only gives 75W), after which I bought RX6400 only to discover what we now both know that not all AMD cards come with encoding support. F me sideways, twice.

The list of what AMD card support different features that I finally discovered, see row Video encoding https://en.wikipedia.org/wiki/List_of_AMD_graphics_processing_units

> perhaps the old RX 580 supported GPU encoding and this new one doesn't,

Yes, exactly, it does. Good to know that VAAPI encoding atleast works.

**Comment by thennicke** [comment]
Just revisiting this issue to state that we got the cause wrong. In fact, the issue was that Fedora has dropped support (since F37) for a lot of VAAPI features due to a patent issue relating to AVC. See the discussion [here](https://www.reddit.com/r/Fedora/comments/xpt34f/fedora_37_drops_vaapi_accelerated_hardware_video/).

I went out and bought myself a new card (RX 6900 XT) only to find it had the same issue, so I did more research and found this out. For anybody using Fedora who wants to make their AMD card work again, run the two commands listed [here](https://rpmfusion.org/Howto/Multimedia) and it'll be fixed. `vainfo` is now showing a lot more features for me.

And I now have a more powerful, and more power efficient graphics card; yay! Can't wait to get back into editing.
